### ISSUE TITLE
Feature Request : Configure to auto add DLC and Updates to game from folder.

### ISSUE THREAD

**Issue by TomTurbine** [Ryujinx/Ryujinx#3265]
#### Overview:

Basically, having the ability to configure Ryujinx so when it detects a game, it will automatically scan predefined areas for potential DLC and Updates to it.

#### Smaller Details: 

Any method to auto scan would be greatly appreciated by many. But some potential ways I can think of would be:

1) Set it up where when it detects a game, it checks any subfolders of the same name for any potential updates/dlc and adds them or..

2) Have Update/DLC Folders setup where when it detects a game, it will automatically generate a folder with that titles ID number and name where people can drag and drop the files to or even right click the game and have them moved that way if need be.

### Why would this feature be useful?

When a game update comes out, that way you no longer had to manually delete an older update and then add the newer one from the menu, just delete the old one and add the new one and it would be handled automatically.

When it comes to people who like to keep backups or update a system with a large library, instead of having to manually add all of them from the menus again, just have the emulator check the stuff when the games are detected or checked.

In both of those situation, just delete any outdated updates and the emulator will automatically replace it with the newer one you put in the folder, any time you added a DLC, the emulator would automatically add it to the game so long as you have it in the correct folder.

**Comment by Neipas09** [comment]
Damn this has been requested for over a year :(

**Comment by TomTurbine** [comment]
Any word on if this feature will ever be added?

**Comment by minhnhattonthat** [comment]
For the time being, I have created a script to scan a specified folder and add the updates/dlcs for ryujinx

https://github.com/minhnhattonthat/ryujinx-tool

**Comment by KruyKaze** [comment]
> For the time being, I have created a script to scan a specified folder and add the updates/dlcs for ryujinx
> 
> https://github.com/minhnhattonthat/ryujinx-tool

Referring to step 2 in your link, is this a windows only solution?
thanks

**Comment by minhnhattonthat** [comment]
> > For the time being, I have created a script to scan a specified folder and add the updates/dlcs for ryujinx
> > https://github.com/minhnhattonthat/ryujinx-tool
> 
> Referring to step 2 in your link, is this a windows only solution? thanks

Yup it's only for windows unfortunately, since I haven't found any multiplatform hactool alternative. Hopefully the Ryujinx team can intergrate this feature in the near future

**Comment by KruyKaze** [comment]
> > > For the time being, I have created a script to scan a specified folder and add the updates/dlcs for ryujinx
> > > https://github.com/minhnhattonthat/ryujinx-tool
> > 
> > 
> > Referring to step 2 in your link, is this a windows only solution? thanks
> 
> Yup it's only for windows unfortunately, since I haven't found any multiplatform hactool alternative. Hopefully the Ryujinx team can intergrate this feature in the near future

hactool itself has a linux release
![image](https://github.com/Ryujinx/Ryujinx/assets/70369/2af4022c-4800-41be-84a3-50c12ab29893)
Very handy for Batocera/Linux users. Any possibility you can adapt your script to allow for this?
Thank you

**Comment by minhnhattonthat** [comment]
@KruyKaze yeah you are right! I have updated the script to be compatible with linux.
If there is any other issue, maybe you can open one on my repo, so we can keep this issue on-topic.

**Comment by KruyKaze** [comment]
> > @KruyKaze yeah you are right! I have updated the script to be compatible with linux. If there is any other issue, maybe you can open one on my repo, so we can keep this issue on-topic.
> 
> hi. I tried to use this on Steam Deck, but no matter where I place the script and files, I always end with following, even running it as a sudo command: " PermissionError: [Errno 13] Permission denied: '/run/media/deck/ae6402c7-3f91-4a4a-b88b-b086545a6073/RyujinxUpdatesScript/ryujinx-tool/hactoolnet' "
> 
> I've tried putting the whole setup in /home/deck/Downloads, in Documents, on my sd card, and even running as sudo, I always get "Errno 13". Any ideas?

Worked fine for me (with updates at least i didn't test DLC) . Thanks @minhnhattonthat 
Make sure you chmod +x your executables @MultiKoopa

**Comment by MultiKoopa** [comment]
> > > @KruyKaze yeah you are right! I have updated the script to be compatible with linux. If there is any other issue, maybe you can open one on my repo, so we can keep this issue on-topic.
> > 
> > 
> > hi. I tried to use this on Steam Deck, but no matter where I place the script and files, I always end with following, even running it as a sudo command: " PermissionError: [Errno 13] Permission denied: '/run/media/deck/ae6402c7-3f91-4a4a-b88b-b086545a6073/RyujinxUpdatesScript/ryujinx-tool/hactoolnet' "
> > I've tried putting the whole setup in /home/deck/Downloads, in Documents, on my sd card, and even running as sudo, I always get "Errno 13". Any ideas?
> 
> Worked fine for me (with updates at least i didn't test DLC) . Thanks @minhnhattonthat Make sure you chmod +x your executables @MultiKoopa

yup that was it. thank you!

**Comment by cddude** [comment]
> @KruyKaze yeah you are right! I have updated the script to be compatible with linux. If there is any other issue, maybe you can open one on my repo, so we can keep this issue on-topic.

Can you please update it to be compatible with MacOS?  Thank you.

**Comment by TomTurbine** [comment]
They REALLY need to integrate this directly into Ryujinx at this point. Would simplify SO much for many people, especially if they have their stuff setup on a portable setup.

This and the lack of controller hotkeys are two of the main things that are missing from it for many people.

**Comment by cddude** [comment]
> They REALLY need to integrate this directly into Ryujinx at this point. Would simplify SO much for many people, especially if they have their stuff setup on a portable setup.
> 
> This and the lack of controller hotkeys are two of the main things that are missing from it for many people.

Absolutely

**Comment by TomTurbine** [comment]
This isn't even something that hard to implement I would think as the emulator can already differentiate between a game and update and dlc.

Why this and configurable hotkeys aren't implemented yet, I don't understand.

**Comment by dcuellar** [comment]
bummer. I was hoping this would have been implemented. Thanks for the tool. I'll try that route instead.

**Comment by dcuellar** [comment]
Xiety, you've reacted with the confused face for the last three comments. Is there something we are missing?

**Comment by TomTurbine** [comment]
@MetrosexualGarbodor

You locked the other one but the PR you referenced even referenced this suggestion as well which predated all the way from April 7, 2022.

**Comment by TomTurbine** [comment]
@gdkchan 

Any word on this feature yet? I know it might be rude to ask but its been about 2 years so worth the while to just see if it has any progress at all.

Having to manually add them in the emulator every time an update is done or manually edit json files for relative directories if you have a portable setup you like to play on multiple PCs might get tedious.

And this and controller hotkeys are the 2 biggest missing features I can think of that isn't directly related to the emulation itself. And was one of the big things that separated Ryujinx from the other emulator for many users.

And getting stuff just referred to being duplicates of this or of another suggestion that is just another version of this doesn't really show any progress made toward them.

Saying "We won't do this" or "We can't do this with the current implementation" or you can do it but have issues or just something would be nice to hear.

Just giving the users a virtual nand to install stuff to that would stay relative to the directory and any updates automatically overwrote the previous one would be nice.

**Comment by MutantAura** [comment]
> Any word on this feature yet? I know it might be rude to ask but its been about 2 years so worth the while to just see if it has any progress at all.

This is a feature _request_. It does not mean that anyone is currently working on such a feature. If they do then the pull request will be linked with this issue.

> Saying "We won't do this" or "We can't do this with the current implementation" or you can do it but have issues or just something would be nice to hear.

The answer is that for every feature someone needs to actually do the work and implement it. Everyone who works on Ryujinx does so under a voluntary basis and as such works on whatever they personally find interesting or useful.

**Comment by TomTurbine** [comment]
Understood, thank you for the response, Do hope this one gets taken up.

**Comment by J-Swift** [comment]
@MutantAura not to impose, but it looks like this PR has been open for 1.5 years and seems to be most of the way there but in a holding pattern? Any idea whats holding it?

**Comment by piplup55** [comment]
> not to impose, but it looks like [this PR](https://github.com/Ryujinx/Ryujinx/pull/4099) has been open for 1.5 years and seems to be most of the way there but in a holding pattern? Any idea whats holding it?

if you look at the tags it says `review/waiting-on-author` which means it's waiting on the author i'm not sure why that is still on the PR i don't see a reason, but the PR would need a rebase at this point

**Comment by J-Swift** [comment]
Just following up on this thread for anyone interested, you can test this on the PR if you are comfortable running the code directly https://github.com/Ryujinx/Ryujinx/pull/7205
